csharp/src/SeedLang.X/PythonParser.cs

```diff
@@ -23,24 +24,25 @@ namespace SeedLang.X {
   // the AST tree based on the predefined rules.
   public sealed class PythonParser {
     // Parses SeedPython source code into the AST tree based on the parse rule.
-    public static AstNode Parse(string source, ParseRule rule) {
-      try {
-        var inputStream = new AntlrInputStream(source);
-        var lexer = new SeedPythonLexer(inputStream);
-        var tokenStream = new CommonTokenStream(lexer);
-        var parser = new SeedPythonParser(tokenStream);
-        var visitor = new PythonVisitor();
-        switch (rule) {
-          case ParseRule.Expression:
-            return visitor.Visit(parser.expr());
-          case ParseRule.Statement:
-            return visitor.Visit(parser.stmt());
-          default:
-            throw new ArgumentException("Unknown parse rule: " + rule);
-        }
-      } catch (Exception) {
-        // TODO: implement parse exception handling.
-        throw;
+    public static AstNode Parse(string source, ParseRule rule, DiagnosticCollection diagnostics) {
+      var inputStream = new AntlrInputStream(source);
+      var lexer = new SeedPythonLexer(inputStream);
+      var tokenStream = new CommonTokenStream(lexer);
+      var parser = new SeedPythonParser(tokenStream);
+
+      // Remove default error listerners which include a console error reporter.
+      parser.RemoveErrorListeners();
+      var errorListener = new SyntaxErrorListener(diagnostics);
+      parser.AddErrorListener(errorListener);
+      var visitor = new PythonVisitor();
+
+      switch (rule) {
+        case ParseRule.Expression:
+          return visitor.Visit(parser.expr());
+        case ParseRule.Statement:
+          return visitor.Visit(parser.stmt());
+        default:
+          throw new ArgumentException("Unknown parse rule: " + rule);

```

Comment:
Do we need to catch this exception (and, others that are thrown by underlying components) in the shell's main loop?

And, if the message string of this exception will be seen by end-users, it should be localized by the Common.Message class. If so, please switch to the Common.Message solution or at least add TODO comments here.

Reply:
This default case wull never happen, because the value is out of the range of the enum values. We can have several options:
1. Use Debug.Assert() and let the program terminate if it really happen (not possible)
2. Use InvalidEnumArgumentException, and let the program terminate.
3. Use InvalidEnumArgumentException, and catch the exception in the shell's main loop. But what we can do is printing the exception. It's as same as the default behavior. 
What's your thought about it?

Reply:
If the case would not happen, either one of your choices is okay I think.